BoxOffice [Step4] Volga

volgaHwang opened the pull request:
@zdodev 

안녕하세요~ 소대!

STEP4 PR을 보내드립니다..!

# 요구사항

- 영화의 상세내용을 확인할 수 있는 화면을 구현합니다.
- 표시할 영화 정보
    - 제목 [✅ ]
    - 포스터 이미지[✅ ]
    - 감독[✅ ]
    - 제작년도[✅ ]
    - 개봉일[✅ ]
    - 상영시간[✅ ]
    - 관람등급[✅ ]
    - 제작국가[✅ ]
    - 장르[✅ ]
    - 배우[✅ ]
- 이미지가 로드되기 전에는 이미지가 로드중임을 알 수 있도록 적절히 처리해주세요 [✅ ]
- 내용이 길어지면 위아래로 스크롤 할 수 있도록 구현해주세요.[✅ ]

## 고민한 점
- 기존에 쓰던 `URLInfo` 을 많이 수정했습니다.
    - 타입 프로퍼티로 `baseURL`을 가져다 사용했었는데 카카오 API에 대한 접근이 필요해진만큼 `baseURL` 대신 `URLComponents`에 `scheme`, `path`, `host`, `query`를 넣어서 만드는 방식으로 구현사항을 변경했습니다.
    - 기존에 `App Key`를 `URLInfo`에서 얻어올수 있게 만들었는데, `Key`에 대한 구현 방식에 대해서, `AppKeys`의 타입 프로퍼티로 가져올수 있게 구현사항을 변경했습니다.
        - 박스오피스 & 카카오 앱 키 모두 `Header` 또는 `Query`에 넣든 `Dictionary` 형태의 상태가 필요하므로 공통적으로 `[String: String]` 형태로 반환하게 구현했습니다.
- `NetworkService`  내부 `fetch()`메소드에 파리미터로 `Header: [String: String]?`을 추가했습니다.
    - 박스 오피스 앱 키가 기존에 쿼리로만 들어갔던 것을 고려하여 뺏던 헤더 부분을 추가했고 URLRequest를 만드는 과정에서 추가하게 변경했습니다.
    - HTTPHeader 정보로 필요한 `Method`에 대해 Enum으로 구현했습니다.
- `DetailMovieInfoViewController` 구현 사항중,`영화 정보`와 `포스터 이미지`에 대해 따로 받아오는 메소드를 구현했습니다.
- `DetailMovieInfoViewController` 에서 `movieCode`, `movieName`프로퍼티를 선언해서 `BoxOfficeListViewController`에서 주입하여 사용하게 구현했습니다.
- 다음화면으로 `movieCode`와 `movieName`이 필요한 만큼 해당 부분에 대해 넘겨 줄 수 있도록 `BoxOfficeItem`에 `movieCode`를 추가했습니다.

## 궁금한 점
- `DetailMovieInfoViewController`로 프로퍼티 정보를 주입해서 넘겨주기 위해서 BoxOfficeItem에 새로운 프로퍼티를 추가해서 구현을 했는데, 이게 객체지향적으로 올바른거 같지 않다고 생각은 하지만, 다른 방식으로 값을 넘겨 주는 것에 대해 방도를 잘 모르겠습니다. (애초에 저장하지 않는 정보를 어떻게.. 넘기지? 라는 생각입니다.)
- `ActivityIndicator`를 `mainStackView` 최 상단에 넣어놓고 `Animating`을 시켜놨는데, 이 부분에 대해서 더 좋은 방법이 있는지 궁금합니다.

volgaHwang committed:
test: NetworkTest Test Case 추가

volgaHwang committed:
test: MockUpTest fetch 변경사항 반영

volgaHwang committed:
feat: NetworkService Fetch 메소드 변경 및 추가 사항 반영

- AppKeys 추가해서 AppKey에 대한 정보를 URLQuery 가 아닌, Dictionary 로 받게 재구현했습니다.
    - 기존의 구조에서 추가하는 방향으로 가게 되면 해당 행위만을 위한 코드가 되고 추상화가 안된다고 판단했습니다.
- `URLInfo` 에서는 `scheme`, `host`, `path`, `defaultQuery`에 대한 내용을 획득해서 `URLComponents`를 구성할수 있도록 구현했습니다.
-

volgaHwang committed:
feat: ImageSearchResult Model 추가

volgaHwang committed:
feat: KAKAO APP KEY 추가 및 BoxOffice App Key 네이밍 변경사항 반영

volgaHwang committed:
feat: NetworkService Fetch Method 변경사항 ViewController에 반영

volgaHwang committed:
feat: 자세한 영화 정보 조회를 위한 뷰컨트롤러 추가

volgaHwang committed:
feat: 자세한 영화 정보 조회를 위한 movieCode 추가

volgaHwang committed:
feat: DetailMoviewInfoViewController 추가 반영

- DetailMoviewInfoViewController 에 속성, 계층, 레이아웃을 구성하는 configure 메소드 틀을 만들었습니다.
- BoxOfficeListViewController에서 property를 넘겨줄수 있게 반영했습니다.

volgaHwang committed:
fix: DetailMovieInfoViewController에 테스트용 print 구문 제거

volgaHwang committed:
feat: DetailMoviewInfoViewController에서 뷰 구성에 필요한 프로퍼티 추가 반영

volgaHwang committed:
fix: 디버그용 print 구문 제거

volgaHwang committed:
feat: DetailMoviewInfoViewController 화면 구성을 위한 configure 메소드 frame 구현

volgaHwang committed:
feat: QueryKeys: 영화 포스터 획득을 위한 추가 Constant 반영

volgaHwang committed:
feat: DetailMovieInfoViewController 에 Network + ActivityIndicator 추가 반영

- Image /  Data Loading을 위한 Network Method 를 추가했습니다.
    - load에서는 정확하게 데이터를 받아오고 매핑시켜주는 과정만을 진행하고, 그 뒤에 update prefix를 사용한 메소드를 통해, Label / Image를 렌더링 시키는 방식으로 구현했습니다.
    - DispatchQueue를 networkMethod에서 사용하면 Depth가 깊어지는 현상때문에 지양했습니다.
- ActivityIndicator를 StackView 최상단에 추가해서 렌더링시켰습니다.

zdodev commented on BoxOffice/Network/NetworkService.swift:
```diff
@@ -8,13 +8,18 @@
 import Foundation
 
 final class NetworkService: NetworkServiceProtocol {
+    private enum HTTPMethod {
+        static let get = "GET"
+    }
```
HTTPMethod enum 타입을 구현해보셨네요!👍

zdodev commented on BoxOffice/Network/NetworkService.swift:
```diff
@@ -82,4 +87,22 @@ final class NetworkService: NetworkServiceProtocol {
         return queryItems
     }
+    
+    private func createHTTPRequest(of url: URL,
```
URLRequest 를 생성하는 메서드를 구현해보셨네요. 좋습니다!👍

zdodev commented on BoxOffice/Scene/DetailMovieInfoViewController.swift:
```diff
@@ -0,0 +1,304 @@
+    private let movieCode: String
+    private let movieName: String
+    
+    private let networkService: NetworkServiceProtocol? = NetworkService()
```
`networkService` 값을 직접 생성 및 대입을 하면 의존성을 주입할 수 없어보입니다.
의존성을 줄이기 위해 구현한 프로토콜을 활용할 수 있도록 수정해 볼 수 있을까요?
그리고 옵셔널로 선언하신 이유가 궁금합니다!

zdodev commented on BoxOffice/Scene/DetailMovieInfoViewController.swift:
```diff
@@ -0,0 +1,304 @@
+        title = movieName
+        view.backgroundColor = .systemBackground
+        
+        posterImageView.contentMode = .scaleAspectFit
```
`contentMode` 를 `scaleAspectFit` 으로 설정하신 이유가 궁금합니다!

zdodev commented on BoxOffice/Scene/DetailMovieInfoViewController.swift:
```diff
@@ -0,0 +1,304 @@
+        posterImageView.contentMode = .scaleAspectFit
+        
+        activityView.translatesAutoresizingMaskIntoConstraints = false
+        activityView.startAnimating()
```
`activityView` 의 animating 실행 조건이 궁금합니다.
`configureAttribute` 메서드를 실행할 때마다 animating 이 실행되어야 하고, 언제 종료하는지에 대한 의문이 들었습니다!

volgaHwang replied:
처음 코드 작성하면서 생각한 `activityView`의 `Animating` 실행 조건은 `viewDidLoad`에서 각 `Sub View`들에 대한 `Attribute`가 지정이 된 시점 이라고 생각을 했습니다.

하지만 목적상 사진을 받아오는 시점에, 사진을 받아오고 있다 라는 명시적 표현을 하기 위한 `View`이므로, `loadMoviewPoster`메소드 에서 네트워킹을 시작하기 직전에 수행 되야 된다고 판단이 듭니다.

zdodev commented on BoxOffice/Scene/DetailMovieInfoViewController.swift:
```diff
@@ -0,0 +1,304 @@
+        ])
+    }
+    
+    private func updateLabel(_ target: MovieInfo) {
+        DispatchQueue.main.async { [weak self] in
```
UI 업데이트를 `main dispatch queue` 에서 수행하기 위한 `DispatchQueue` 구문일까요?
UI 를 업데이트하는 메서드가 늘어날 때마다 해당 메서드에 `DispatchQueue` 를 작성해주어야 할 것 같습니다!
이 부분을 비동기를 수행하는 쪽에서 처리하도록 하는 것은 어떨까요? 의견이 궁금합니다!

volgaHwang replied:
말씀주신대로 main Dispatch Queue에서 수행하기 위한 구문이 맞습니다.

지금은 Label만 있지만, 따로 메소드를 추가해야 할 경우 메소드 내에 DispatchQueue를 작성해야 하는 문제가 있는 것이 맞고,
비동기적으로 데이터를 획득 한 이후 updateUI를 진입하는 시점에 DispatchQueue 를 넣어서 UI를 업데이트 할수 있도록 수정했습니다!

zdodev commented on BoxOffice/Scene/DetailMovieInfoViewController.swift:
```diff
@@ -0,0 +1,304 @@
+        if let data = try? Data(contentsOf: url) {
+            if let image = UIImage(data: data) {
+                DispatchQueue.main.async { [weak self] in
+                    self?.activityView.stopAnimating()
```
만약 네트워크 요청이 실패하거나 에러가 발생하는 경우에 `activityView` 가 계속 동작하고 있을 것 같습니다!
예외의 경우에도 `activityView` 가 종료될 수 있도록 해보는 것은 어떨까요?

zdodev commented on BoxOfficeTests/NetworkTest/NetworkTest.swift:
```diff
@@ -86,6 +90,23 @@ final class NetworkTest: XCTestCase {
-
-
+    
+    func test_images_잘_받아_오나() {
```
이미지 요청에 대한 네트워 테스트를 작성해보셨네요.👍

zdodev commented on BoxOfficeTests/NetworkTest/NetworkTest.swift:
```diff
@@ -86,6 +90,23 @@ final class NetworkTest: XCTestCase {
+            case .success(let success):
+                print(success)
+                expects.fulfill()
+            case .failure(let failure):
```
warning 이 발생하지 않도록 수정해보는 것은 어떨까요?

zdodev reviewed (changes requested):
안녕하세요. @volgaHwang !

Step 4 구현 잘 진행해주셨습니다!
적은 코드의 작성으로 기능을 잘 추가해주셨습니다.👍

질문에 대한 제 생각을 적어보았습니다.

❓ DetailMovieInfoViewController로 프로퍼티 정보를 주입해서 넘겨주기 위해서 BoxOfficeItem에 새로운 프로퍼티를 추가해서 구현을 했는데, 이게 객체지향적으로 올바른거 같지 않다고 생각은 하지만, 다른 방식으로 값을 넘겨 주는 것에 대해 방도를 잘 모르겠습니다. (애초에 저장하지 않는 정보를 어떻게.. 넘기지? 라는 생각입니다.)

- 지금도 잘 전달해주셨습니다! 더 개선해본다면 각각의 프로퍼티로 전달하기 보다는 구조체에 담아서 전달하는 방식을 생각해볼 수 있을 것 같습니다.
- 모든 데이터를 저장하거나, 사용자에게 입력받지 않는 부분도 있어서 뷰 컨트롤러 사이, 화면 사이에 데이터 전달은 자주 사용합니다. 다양한 데이터 전달 방식을 고민해보면 좋을 것 같습니다!

❓ ActivityIndicator를 mainStackView 최 상단에 넣어놓고 Animating을 시켜놨는데, 이 부분에 대해서 더 좋은 방법이 있는지 궁금합니다.
- 이 부분은 앱을 구현하는 기획에 따라 다를 것 같습니다. 지금도 잘 구현해주셨습니다!
- 다음 링크를 참고사항으로 확인해보면 좋을 것 같습니다!
- [progress-indicators](https://developer.apple.com/design/human-interface-guidelines/components/status/progress-indicators/)

그리고 몇 가지 궁금증을 남겨보았습니다!
확인 부탁드립니다~

volgaHwang committed:
test: XCTFail 발생 시 error Message 출력하도록 반영

volgaHwang committed:
refactor: 의존성 주입 반영

- `DetailMovieInfoViewController` 에서 사용하는 networkService 관련하여 BoxOfficeListViewController에서 주입하도록 변경
- networkServiceProtocol 이 Optional 한 값을 가지는게 의미가 없다고 판단하여 Non-Optional 하게 변경했습니다.

volgaHwang committed:
refactor: DetailMovieInfoViewController

-updateUI를 `loadDetailInfo` 메소드에서 비동기적으로 데이터를 획득한 이후에 `DispatchQueue`를 사용하도록 구현했습니다.
- `activityView`를 에러나, 네트워킹 실패 상황에서도 stopAnimating할 수 있도록 따로 메소드를 작성했고, `success`, `failure`상황 모두 stop을 할 수 있게 구현했습니다.

volgaHwang commented:
@zdodev 

안녕하세요 소대!

이번에도 좋은 리뷰 감사드립니다..!
말씀해주셨던 리뷰 내용들에 대한 부분을 리팩토링 진행했습니다~!

추가적으로 질문 주신 부분들에 대하여 제 생각을 전달드립니다!

> UI 업데이트를 main dispatch queue 에서 수행하기 위한 DispatchQueue 구문일까요? <br>UI 를 업데이트하는 메서드가 늘어날 때마다 해당 메서드에 DispatchQueue 를 작성해주어야 할 것 같습니다!<br>이 부분을 비동기를 수행하는 쪽에서 처리하도록 하는 것은 어떨까요? 의견이 궁금합니다!

- 말씀주신대로 main Dispatch Queue에서 수행하기 위한 구문이 맞습니다.

- 지금은 Label만 있지만, 따로 메소드를 추가해야 할 경우 메소드 내에 DispatchQueue를 작성해야 하는 문제가 있는 것이 맞고, 비동기적으로 데이터를 획득 한 이후 updateUI를 진입하는 시점에 DispatchQueue 를 넣어서 UI를 업데이트 할수 있도록 수정했습니다!

> activityView 의 animating 실행 조건이 궁금합니다.<br>configureAttribute 메서드를 실행할 때마다 animating 이 실행되어야 하고, 언제 종료하는지에 대한 의문이 들었습니다!

- 처음 코드 작성하면서 생각한 activityView의 Animating 실행 조건은 viewDidLoad에서 각 Sub View들에 대한 Attribute가 지정이 된 시점 이라고 생각을 했습니다.

- 하지만 목적상 사진을 받아오는 시점에, 사진을 받아오고 있다 라는 명시적 표현을 하기 위한 View이므로, loadMoviewPoster메소드 에서 네트워킹을 시작하기 직전에 수행 되야 된다고 판단이 듭니다.

> networkService 값을 직접 생성 및 대입을 하면 의존성을 주입할 수 없어보입니다.<br>의존성을 줄이기 위해 구현한 프로토콜을 활용할 수 있도록 수정해 볼 수 있을까요?<br>그리고 옵셔널로 선언하신 이유가 궁금합니다!

- 최악의 답변이 될 것 같습니다.. 네트워크 통신이 되는지 확인하는데 있어서 처음에 `NetworkService` 클래스를 주입을 제대로 하지 않아도 동작을 할 수 있게끔 비슷하게 구현을 해왔었는데 무심코 그냥 그대로 쓴 것 같습니다.
- 의존성 주입을 위해서 프로토콜을 활용해서 주입 할 수 있도록 재 선언했습니다!

> contentMode 를 scaleAspectFit 으로 설정하신 이유가 궁금합니다!

- 따로 명세에는 그림에 대해서 잘려야한다거나 하는 명세사항은 없었습니다만, 영화 포스터를 보여주는 것이 목표인 만큼 영화 포스터 내용이 잘려서는 안된다고 판단했습니다.
- `contentMode`에는 대표적으로 세가지 케이스가 있는데, `scaleAspectFill`의 경우에는 이미지가 잘릴수 있고, `scaleToFill`의 경우에는 비율이 깨지는 문제점이 있기 때문에 `scaleAspectFit`이 영화 포스터 정보를 보여주는 데에 있어 더 적합하다고 판단하였습니다!
-

zdodev commented on BoxOffice/Scene/BoxOfficeListViewController.swift:
```diff
@@ -37,7 +37,7 @@ class BoxOfficeListViewController: UIViewController {
                                                                BoxOfficeItem>?
     private var items: [BoxOfficeItem] = []
-    private let networkService: NetworkServiceProtocol? = NetworkService()
+    private let networkService: NetworkServiceProtocol = NetworkService()
```
좋습니다!👍

zdodev commented on BoxOffice/Scene/DetailMovieInfoViewController.swift:
```diff
@@ -283,24 +283,31 @@ extension DetailMovieInfoViewController {
     }
     
+    private func stopAnimateLoadingImage() {
+        DispatchQueue.main.async { [weak self] in
+            self?.activityView.stopAnimating()
+        }
```
`DispatchQueue` 구문을 `NetworkService` 의 `completion` 부분에 추가하는 것도 괜찮을 것 같습니다!

zdodev reviewed (approved):
안녕하세요. @volgaHwang !

코멘트 확인해보았습니다!
질문에 대한 생각을 잘 적어주셔서 좋았습니다!👍

코멘트만 확인하고 넘어가면 될 것 같습니다!
고생하셨습니다~🎉
